src/Storages/KeyDescriptionMonotonicityChecker.cpp

```diff
@@ -0,0 +1,247 @@
+                }
+                else if (function_node.getArgumentAt(1).isConstant())
+                {
+                    buildRPNFunctionListImpl(function_node.getArgumentAt(0), key_description, out_key_column_type, function_list);
+                }
+                else
+                {
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS,
+                                    "A function with 2 arguments must contain at least one constant argument to be considered monotonic");
+                }
+            }
+            else
+            {
+                buildRPNFunctionListImpl(function_node.getArgumentAt(0), key_description, out_key_column_type, function_list);
+            }
+        }
+        else
+        {
+            // assuming the column is always part of expression
+
+            auto node_name = node.getColumnName();
+            auto key_column = key_description.expression->getSampleBlock().getByName(node_name);
+            out_key_column_type = key_column.type;
+        }
+    }
+
+    Field applyFunctionForField(
+        const FunctionBasePtr & func,
+        const DataTypePtr & arg_type,
+        const Field & arg_value)
+    {
+        ColumnsWithTypeAndName columns
+            {
+                { arg_type->createColumnConst(1, arg_value), arg_type, "x" },
+            };
+
+        auto col = func->execute(columns, func->getResultType(), 1);
+        return (*col)[0];
+    }
+
+    FieldRef applyFunction(const FunctionBasePtr & func, const DataTypePtr & current_type, const FieldRef & field)
+    {
+        /// Fallback for fields without block reference.
+        if (field.isExplicit())
+            return applyFunctionForField(func, current_type, field);
+
+        String result_name = "_" + func->getName() + "_" + toString(field.column_idx);
+        const auto & columns = field.columns;
+        size_t result_idx = columns->size();
+
+        for (size_t i = 0; i < result_idx; ++i)
+        {
+            if ((*columns)[i].name == result_name)
+                result_idx = i;
+        }
+
+        if (result_idx == columns->size())
+        {
+            ColumnsWithTypeAndName args{(*columns)[field.column_idx]};
+            field.columns->emplace_back(ColumnWithTypeAndName {nullptr, func->getResultType(), result_name});
+            (*columns)[result_idx].column = func->execute(args, (*columns)[result_idx].type, columns->front().column->size());
+        }
+
+        return {field.columns, field.row_idx, result_idx};
+    }
+
+    IFunction::Monotonicity applyMonotonicFunctionsChainToRange(
+        Range key_range,
+        const std::vector<FunctionBasePtr> & functions,
+        DataTypePtr current_type)
+    {
+        // default to monotonically increasing, satisfies cases where function list is empty
+        IFunction::Monotonicity monotonicity {true, true, true, true};
```

Comment:
Original code has a "range single point check" which did not seem relevant to our case, so I removed it.

Sometimes an AST might not contain a function (i.e PARTITION BY column_a). In that case, it's monotonic, that's why this defaults to true.